server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java

```diff
@@ -104,26 +121,41 @@ public long ramBytesUsed() {
 
     @Override
     public StreamInput streamInput() throws IOException {
-        return delegate.streamInput();
+        assert refCount() > 0;
+        return new BytesReferenceStreamInput(this) {
+            @Override
+            public ReleasableBytesReference readUnsafeBytesReference() throws IOException {
+                final int len = readArraySize();
+                // instead of reading the bytes from a stream we just create a slice of the underlying bytes
+                final ReleasableBytesReference result = ((ReleasableBytesReference) bytesReference).retainedSlice(offset(), len);
```

Comment:
Maybe avoid the casting?

```suggestion
                final ReleasableBytesReference result = ReleasableBytesReference.this.retainedSlice(offset(), len);
```

Reply:
++